File: Examples/CSExamples/JustMock.NonElevatedExamples/AdvancedUsage/MockingDbContextAsync.cs
using System;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.EntityFrameworkCore;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Telerik.JustMock;

namespace JustMock.NonElevatedExamples.AdvancedUsage.MockingDbContextAsync
{
    /// <summary>
    /// Virtual EF Core members can be arranged with JustMock Lite while asynchronous
    /// healthcare queries and persistence use the EF Core InMemory provider.
    /// </summary>
    [TestClass]
    public class MockingDbContextAsync_Tests
    {
        [TestMethod]
        public async Task ShouldQueryArrangedVirtualPatientSetAsynchronously()
        {
            using (var backingContext = AsyncHealthcareContext.CreateInMemory("MockingDbContextAsync.Query"))
            {
                backingContext.Patients.AddRange(
                    new AsyncPatient { Id = 1, Name = "Olivia Carter", Department = "Cardiology", DoctorId = 10, IsActive = true },
                    new AsyncPatient { Id = 2, Name = "Liam Turner", Department = "Cardiology", DoctorId = 10, IsActive = false },
                    new AsyncPatient { Id = 3, Name = "Maya Patel", Department = "Cardiology", DoctorId = 20, IsActive = true },
                    new AsyncPatient { Id = 4, Name = "Noah Williams", Department = "Pediatrics", DoctorId = 30, IsActive = true });
                await backingContext.SaveChangesAsync();

                var context = Mock.Create<AsyncHealthcareContext>();
                Mock.Arrange(() => context.Patients).Returns(backingContext.Patients);

                var actual = await new AsyncPatientDirectory(context).FindActiveNamesAsync("Cardiology");

                CollectionAssert.AreEqual(new[] { "Maya Patel", "Olivia Carter" }, actual);
            }
        }

        [TestMethod]
        public async Task ShouldSavePatientThroughArrangedVirtualContextAsynchronously()
        {
            using (var backingContext = AsyncHealthcareContext.CreateInMemory("MockingDbContextAsync.Save"))
            {
                var context = Mock.Create<AsyncHealthcareContext>();
                Mock.Arrange(() => context.Patients).Returns(backingContext.Patients);
                Mock.Arrange(() => context.SaveChangesAsync(Arg.IsAny<CancellationToken>()))
                    .ReturnsAsync(1)
                    .MustBeCalled();

                var result = await new AsyncPatientWriter(context).AddAsync(new AsyncPatient
                {
                    Id = 5,
                    Name = "Ava Brooks",
                    Department = "Pediatrics",
                    DoctorId = 30,
                    IsActive = true
                });

                Assert.AreEqual(1, result);
                Assert.AreEqual(1, backingContext.Patients.Local.Count);
                Mock.Assert(context);
            }
        }

        [TestMethod]
        public async Task ShouldVerifyAsyncPatientSaveOccurrence()
        {
            using (var backingContext = AsyncHealthcareContext.CreateInMemory("MockingDbContextAsync.Occurrence"))
            {
                var context = Mock.Create<AsyncHealthcareContext>();
                Mock.Arrange(() => context.Patients).Returns(backingContext.Patients);
                Mock.Arrange(() => context.SaveChangesAsync(Arg.IsAny<CancellationToken>()))
                    .ReturnsAsync(1);

                var writer = new AsyncPatientWriter(context);
                await writer.AddAsync(new AsyncPatient
                {
                    Id = 6,
                    Name = "Ethan Clark",
                    Department = "Cardiology",
                    DoctorId = 10,
                    IsActive = true
                });
                await writer.AddAsync(new AsyncPatient
                {
                    Id = 7,
                    Name = "Sofia Green",
                    Department = "Neurology",
                    DoctorId = 20,
                    IsActive = true
                });

                Mock.Assert(
                    () => context.SaveChangesAsync(Arg.IsAny<CancellationToken>()),
                    Occurs.Exactly(2));
            }
        }
    }

    public class AsyncHealthcareContext : DbContext
    {
        public AsyncHealthcareContext()
        {
        }

        public AsyncHealthcareContext(DbContextOptions<AsyncHealthcareContext> options)
            : base(options)
        {
        }

        public virtual DbSet<AsyncPatient> Patients { get; set; }

        public virtual DbSet<AsyncDoctor> Doctors { get; set; }

        public static AsyncHealthcareContext CreateInMemory(string databaseName)
        {
            var options = new DbContextOptionsBuilder<AsyncHealthcareContext>()
                .UseInMemoryDatabase(databaseName)
                .Options;
            var context = new AsyncHealthcareContext(options);

            context.Database.EnsureDeleted();
            return context;
        }
    }

    public class AsyncPatientDirectory
    {
        private readonly AsyncHealthcareContext context;

        public AsyncPatientDirectory(AsyncHealthcareContext context)
        {
            this.context = context;
        }

        public async Task<string[]> FindActiveNamesAsync(
            string department,
            CancellationToken cancellationToken = default)
        {
            return await context.Patients
                .AsNoTracking()
                .Where(patient => patient.Department == department && patient.IsActive)
                .OrderBy(patient => patient.Name)
                .Select(patient => patient.Name)
                .ToArrayAsync(cancellationToken);
        }
    }

    public class AsyncPatientWriter
    {
        private readonly AsyncHealthcareContext context;

        public AsyncPatientWriter(AsyncHealthcareContext context)
        {
            this.context = context;
        }

        public async Task<int> AddAsync(
            AsyncPatient patient,
            CancellationToken cancellationToken = default)
        {
            await context.Patients.AddAsync(patient, cancellationToken);
            return await context.SaveChangesAsync(cancellationToken);
        }
    }

    public class AsyncPatient
    {
        public int Id { get; set; }

        public string Name { get; set; }

        public string Department { get; set; }

        public int DoctorId { get; set; }

        public bool IsActive { get; set; }
    }

    public class AsyncDoctor
    {
        public int Id { get; set; }

        public string Name { get; set; }

        public string Specialty { get; set; }
    }
}

File: Examples/VBExamples/JustMock.NonElevatedExamples/AdvancedUsage/MockingDbContextAsync.vb
Imports System
Imports System.Linq
Imports System.Threading
Imports System.Threading.Tasks
Imports Microsoft.EntityFrameworkCore
Imports Microsoft.VisualStudio.TestTools.UnitTesting
Imports Telerik.JustMock

Namespace JustMock.NonElevatedExamples.AdvancedUsage.MockingDbContextAsync
    ''' <summary>
    ''' Overridable EF Core members can be arranged with JustMock Lite while asynchronous
    ''' healthcare queries and persistence use the EF Core InMemory provider.
    ''' </summary>
    <TestClass>
    Public Class MockingDbContextAsync_Tests
        <TestMethod>
        Public Async Function ShouldQueryArrangedVirtualPatientSetAsynchronously() As Task
            Using backingContext = AsyncHealthcareContext.CreateInMemory("MockingDbContextAsync.Query")
                backingContext.Patients.AddRange(
                    New AsyncPatient With {.Id = 1, .Name = "Olivia Carter", .Department = "Cardiology", .DoctorId = 10, .IsActive = True},
                    New AsyncPatient With {.Id = 2, .Name = "Liam Turner", .Department = "Cardiology", .DoctorId = 10, .IsActive = False},
                    New AsyncPatient With {.Id = 3, .Name = "Maya Patel", .Department = "Cardiology", .DoctorId = 20, .IsActive = True},
                    New AsyncPatient With {.Id = 4, .Name = "Noah Williams", .Department = "Pediatrics", .DoctorId = 30, .IsActive = True})
                Await backingContext.SaveChangesAsync()

                Dim context = Mock.Create(Of AsyncHealthcareContext)()
                Mock.Arrange(Function() context.Patients).Returns(backingContext.Patients)

                Dim actual = Await New AsyncPatientDirectory(context).FindActiveNamesAsync("Cardiology")

                CollectionAssert.AreEqual(New String() {"Maya Patel", "Olivia Carter"}, actual)
            End Using
        End Function

        <TestMethod>
        Public Async Function ShouldSavePatientThroughArrangedVirtualContextAsynchronously() As Task
            Using backingContext = AsyncHealthcareContext.CreateInMemory("MockingDbContextAsync.Save")
                Dim context = Mock.Create(Of AsyncHealthcareContext)()
                Mock.Arrange(Function() context.Patients).Returns(backingContext.Patients)
                Mock.Arrange(Function() context.SaveChangesAsync(Arg.IsAny(Of CancellationToken)())).
                    ReturnsAsync(1).
                    MustBeCalled()

                Dim result = Await New AsyncPatientWriter(context).AddAsync(New AsyncPatient With {
                    .Id = 5,
                    .Name = "Ava Brooks",
                    .Department = "Pediatrics",
                    .DoctorId = 30,
                    .IsActive = True
                })

                Assert.AreEqual(1, result)
                Assert.AreEqual(1, backingContext.Patients.Local.Count)
                Mock.Assert(context)
            End Using
        End Function

        <TestMethod>
        Public Async Function ShouldVerifyAsyncPatientSaveOccurrence() As Task
            Using backingContext = AsyncHealthcareContext.CreateInMemory("MockingDbContextAsync.Occurrence")
                Dim context = Mock.Create(Of AsyncHealthcareContext)()
                Mock.Arrange(Function() context.Patients).Returns(backingContext.Patients)
                Mock.Arrange(Function() context.SaveChangesAsync(Arg.IsAny(Of CancellationToken)())).
                    ReturnsAsync(1)

                Dim writer = New AsyncPatientWriter(context)
                Await writer.AddAsync(New AsyncPatient With {
                    .Id = 6,
                    .Name = "Ethan Clark",
                    .Department = "Cardiology",
                    .DoctorId = 10,
                    .IsActive = True
                })
                Await writer.AddAsync(New AsyncPatient With {
                    .Id = 7,
                    .Name = "Sofia Green",
                    .Department = "Neurology",
                    .DoctorId = 20,
                    .IsActive = True
                })

                Mock.Assert(
                    Function() context.SaveChangesAsync(Arg.IsAny(Of CancellationToken)()),
                    Occurs.Exactly(2))
            End Using
        End Function
    End Class

    Public Class AsyncHealthcareContext
        Inherits DbContext

        Public Sub New()
        End Sub

        Public Sub New(options As DbContextOptions(Of AsyncHealthcareContext))
            MyBase.New(options)
        End Sub

        Public Overridable Property Patients As DbSet(Of AsyncPatient)

        Public Overridable Property Doctors As DbSet(Of AsyncDoctor)

        Public Shared Function CreateInMemory(databaseName As String) As AsyncHealthcareContext
            Dim options = New DbContextOptionsBuilder(Of AsyncHealthcareContext)().
                UseInMemoryDatabase(databaseName).
                Options
            Dim context = New AsyncHealthcareContext(options)

            context.Database.EnsureDeleted()
            Return context
        End Function
    End Class

    Public Class AsyncPatientDirectory
        Private ReadOnly context As AsyncHealthcareContext

        Public Sub New(context As AsyncHealthcareContext)
            Me.context = context
        End Sub

        Public Async Function FindActiveNamesAsync(
            department As String,
            Optional cancellationToken As CancellationToken = Nothing) As Task(Of String())

            Return Await context.Patients.
                AsNoTracking().
                Where(Function(patient) patient.Department = department AndAlso patient.IsActive).
                OrderBy(Function(patient) patient.Name).
                Select(Function(patient) patient.Name).
                ToArrayAsync(cancellationToken)
        End Function
    End Class

    Public Class AsyncPatientWriter
        Private ReadOnly context As AsyncHealthcareContext

        Public Sub New(context As AsyncHealthcareContext)
            Me.context = context
        End Sub

        Public Async Function AddAsync(
            patient As AsyncPatient,
            Optional cancellationToken As CancellationToken = Nothing) As Task(Of Integer)

            Await context.Patients.AddAsync(patient, cancellationToken)
            Return Await context.SaveChangesAsync(cancellationToken)
        End Function
    End Class

    Public Class AsyncPatient
        Public Property Id As Integer
        Public Property Name As String
        Public Property Department As String
        Public Property DoctorId As Integer
        Public Property IsActive As Boolean
    End Class

    Public Class AsyncDoctor
        Public Property Id As Integer
        Public Property Name As String
        Public Property Specialty As String
    End Class
End Namespace
